rust-overlay.nix

```diff
@@ -458,8 +457,8 @@ let
             inherit (manifest) targetComponentsList;
             extensions = componentNames ++ extensions;
             targets = [
-              (toRustTarget self.stdenv.hostPlatform) # Build script requires host std.
-              (toRustTarget self.stdenv.targetPlatform)
+              (toRustTarget self.stdenv.buildPlatform) # Build script requires build platform std.
+              (toRustTarget self.stdenv.hostPlatform)
```

Comment:
No, `hostPlatform` is _not_ where the compiler runs, it is where the resulting derivation (that we are building with Rust) is supposed to run.

Reply:
And in the example, it should be `build = x86_64`, `host=aarch64` since you are producing binaries to run on `aarch64`

Reply:
> it is where the resulting derivation is supposed to run.

We are building the derivation of compiler it self. So it is where the rustc runs.

> And in the example, it should be build = x86_64, host = aarch64

Again, I'm talking about the derivation of rustc. rustc is running on x86_64, so rustc's hostPlatform is x86_64. You can check it by typing `(import <nixpkgs> { overlays = [ (import ./.) ]; crossSystem = "aarch64-linux"; }).buildPackages.rust-bin.stable.latest.rustc.stdenv.hostPlatform` (or targetPlatform, buildPlatform) in `nix repl`.

Reply:
That is because you are (incorrectly) using `buildPackages`.

```
(import <nixpkgs> { overlays = [ (import ./.) ]; crossSystem = "aarch64-linux"; }).rust-bin.stable.latest.rustc.stdenv.hostPlatform.system
"aarch64-linux"
```

Reply:
You should know that compiler itself is specified in `nativeBuildInputs`, which has -1 level (compiler's hostPlatform is outputs' buildPlatform, compiler's targetPlatform is outputs' hostPlatform). `nixpkgs` has a mechanism called package spliting to auto-inject `buildPackages` for `nativeBuildInputs` so that normally you should not manually use it.

But there is a bug that `mkShell` in `nix-shell` doesn't do these injection https://github.com/NixOS/nixpkgs/issues/49526 .
So you should use
```
mkShell {
  nativeBuildInputs = [ buildPackages.rustc ];
}
```
or
```
callPackages ({ mkShell, rustc }: mkShell {
  nativeBuildInputs = [ rustc ];
}) {}
```
These two way produce the same result.

Without `buildPackages` or relying on package splitting, `pkgsCross.rustc` is the same as `pkgsCross.hello`, which contains aarch64 binaries. You never expect your compiler itself is in aarch64.

Reply:
Here we **ARE** packaging the compiler itself. Compiler's `targetPlaform` is just output's `hostPlatform`, which is not avoidable.

I reference these info from nix wiki https://nixos.wiki/wiki/Cross_Compiling , it also explain that `callPackage` should be used in nix-shell to get the expected cross-compiler.

Reply:
Yeah, sorry I reverted these changes, they are wrong and does not contribute to fixing the original problem